lxqt-config-appearance works in Wayland server.

selairi opened the pull request:
Requires: [https://github.com/lxqt/liblxqt/pull/286](https://github.com/lxqt/liblxqt/pull/286)
Gtk 3 needs gsettings instead xsettings in Wayland server.

luis-pereira commented:
I'm currently not using GTK3 nor Wayland. I'm not the best reviewer to this PR.

tsujan commented:
I might not find time to test it under Weston soon but it'll be good to have it in 0.17.

@stefonarch, could you test the PR for setting GTK3 appearance under LXQt (X11)? If not, just tell me.

tsujan commented:
BTW, it should NOT require https://github.com/lxqt/liblxqt/pull/286 anymore.

stefonarch commented:
> BTW, it should NOT require [lxqt/liblxqt#286](https://github.com/lxqt/liblxqt/pull/286) anymore.

No, it _does_ need it, otherwise
```
...
[ 95%] Building CXX object src/CMakeFiles/lxqt-config.dir/mainwindow.cpp.o
[ 95%] Building CXX object lxqt-config-appearance/CMakeFiles/lxqt-config-appearance.dir/main.cpp.o
/home/stef/git/lxqt-config-gtk3/lxqt-config-appearance/main.cpp:33:10: fatal error: LXQt/lxqtplatform.h: File o directory non esistente
   33 | #include <LXQt/lxqtplatform.h>
      |          ^~~~~~~~~~~~~~~~~~~~~
```
Will make some tests later today.

tsujan commented on lxqt-config-appearance/main.cpp:
```diff
@@ -30,6 +30,8 @@
 #include <LXQt/ConfigDialog>
 #include <QCommandLineParser>
+#include <LXQt/lxqtplatform.h>
+
```
Please remove the above header.

selairi replied:
Ok

tsujan commented:
@stefonarch, it can be compiled without that line.

stefonarch commented:
Maybe I'm not the right person to test this, but I see that only gtk2 settings (for example breeze-dark and gnome-disk-utility) are applied, gtk3 settings had no effect. But it was like that also before.

tsujan commented:
You may have found a problem.

tsujan commented:
BTW, for testing GTK3 settings, `gtk3-widget-factory` is the best and simplest tool. (In very old times, I made GTK3 themes and tested them with it; my themes were broken by Gnome devs so many times and I stopped using GTK3.)

selairi commented:
Hi, I have just updated the code:
- GSettings is used instead xsettings. I added more GTK options and commented deprecated ones.
- lxqtplatform.h has been removed.
- Cursor themes are applied on Wayland, but restart is required. 
On Wayland, Qt reads XCURSOR_THEME environment variable at start or use default cursor (see file qwaylandinputdevice.cpp from qtwayland module). I can not find a way to update the cursor theme on Qt.

I have added a warning dialog, when restart is required. I have found that restart is required on X11, also. If you change cursor theme on X11, you will find that some windows don't update the cursor theme, example, the Openbox title bar.

Thanks

selairi commented:
@stefonarch

> Maybe I'm not the right person to test this, but I see that only gtk2 settings (for example breeze-dark and gnome-disk-utility) are applied, gtk3 settings had no effect. But it was like that also before.

On X11 xsettingsd is used to update themes. xsettingsd can not apply different settings to GTK2 and GTK3. The GTK2 settings are applied by default. If you restart the LXQt session, the themes will be Ok.

tsujan reviewed (changes requested):
Worked under Weston after manual editing. The GTK theme didn't change on the fly but I don't think we need to worry about that. The user should restart his GTK apps to see the effect and, IMO, that's enough. But:

 1. It cannot be compiled in its current form because `#include <LXQt/lxqtplatform.h>` is still there.
 2. I don't think consulting `gsettings` under Wayland is a good idea for getting the cursor theme. `gsettings` is for Gnome, not for us. I don't have another solution but I think we shouldn't rely on `gsettings` here.

selairi commented:
@tsujan 
> Worked under Weston after manual editing. The GTK theme didn't change on the fly but I don't think we need to worry about that. The user should restart his GTK apps to see the effect and, IMO, that's enough.

You have to set the environment variables for LXQt and Qt theme changes should work. GTK3 theme should be changed on the fly, too. Qt cursor theme is not changed on the fly, because Qt library reads the cursor theme on start.

Please, try:

````
#!/bin/sh

contains()
{
    local str="$1" substr="$2"
    [ "$str" = "$substr" -o -z "${str##$substr:*}" -o -z "${str##*:$substr:*}" -o -z "${str%%*:$substr}" ]
}

if [ -z "$XDG_DATA_HOME" ]; then
    export XDG_DATA_HOME="$HOME/.local/share"
fi

if [ -z "$XDG_CONFIG_HOME" ]; then
    export XDG_CONFIG_HOME="$HOME/.config"
fi

if [ -z "$XDG_DATA_DIRS" ]; then
    XDG_DATA_DIRS="$XDG_DATA_HOME:/usr/local/share:/usr/share"
else
    if ! contains "$XDG_DATA_DIRS" "/usr/local/share"; then
        XDG_DATA_DIRS="$XDG_DATA_DIRS:/usr/local/share"
    fi
fi
export XDG_DATA_DIRS

if [ -z "$XDG_CONFIG_DIRS" ]; then
    export XDG_CONFIG_DIRS="/etc:/etc/xdg:/usr/share"
else
    if ! contains "$XDG_CONFIG_DIRS" '/etc/xdg'; then
        XDG_CONFIG_DIRS="$XDG_CONFIG_DIRS:/etc/xdg"
    fi
fi

if [ -z "$XDG_CACHE_HOME" ]; then
    export XDG_CACHE_HOME="$HOME/.cache"
fi

# Ensure the existence of the 'Desktop' folder
if [ -e "$XDG_CONFIG_HOME/user-dirs.dirs" ]; then
    . "$XDG_CONFIG_HOME/user-dirs.dirs"
else
    XDG_DESKTOP_DIR="$HOME/Desktop"
fi
mkdir -p "$XDG_DESKTOP_DIR"

export XDG_MENU_PREFIX="lxqt-"

export XDG_CURRENT_DESKTOP="LXQt"

export QTXDG_START_DETACH_TRULY=0

export QT_QPA_PLATFORM=wayland
export GDK_BACKEND=wayland
export QT_STYLE_OVERRIDE=fusion
export SAL_USE_VCLPLUGIN=gtk3
export QT_PLATFORM_PLUGIN=lxqt
export QT_QPA_PLATFORMTHEME=lxqt
export MOZ_ENABLE_WAYLAND=1
# Launch DBus if needed
if [ -z "$DBUS_SESSION_BUS_ADDRESS" ]; then
    if [ -z "$XDG_RUNTIME_DIR" ] || ! [ -S "$XDG_RUNTIME_DIR/bus" ] || ! [ -O "$XDG_RUNTIME_DIR/bus" ]; then
        eval "$(dbus-launch --sh-syntax --exit-with-session)" || echo "startlxqt: error executing dbus-launch" >&2
    fi
fi
# Uncomment your Wayland server
weston
# sway
# wayfire
````

> 
>     1. It cannot be compiled in its current form because `#include <LXQt/lxqtplatform.h>` is still there.
> 

Removed. Sorry.

>     2. I don't think consulting `gsettings` under Wayland is a good idea for getting the cursor theme. `gsettings` is for Gnome, not for us. I don't have another solution but I think we shouldn't rely on `gsettings` here.

This is a good point. gsettings should be used be lxqt-config-appearance in order to set and get GTK settings. GTK3 doesn't read its own config files in Wayland session. GTK3 reads gsettings values. Get default mouse theme in Wayland is easy with gsettings, then I wrote the code.

If gsettings is not used, then a function to read the cursor theme from XDG spec must be written:
[https://wiki.archlinux.org/index.php/Cursor_themes#XDG_specification](https://wiki.archlinux.org/index.php/Cursor_themes#XDG_specification)
"/home/user/.icons/default/index.theme" file must be parsed. If "/home/user/.icons/default/index.theme" doesn't exist, then "/usr/share/icons/default/index.theme" must be parsed.
I could write this function if you want.

Thanks for your suggestions.

tsujan commented:
> gsettings should be used be lxqt-config-appearance in order to set and get GTK settings.

Yes, that's reasonable.

> "/home/user/.icons/default/index.theme" file must be parsed....

`lxqt-config-appearance` already does that; doesn't it? I have `~/.icons/default/index.theme`; it starts with: "# Written by lxqt-config-appearance".

selairi commented:
> > "/home/user/.icons/default/index.theme" file must be parsed....
> 
> `lxqt-config-appearance` already does that; doesn't it? I have `~/.icons/default/index.theme`; it starts with: "# Written by lxqt-config-appearance".

Yes and no. As far as I know `~/.icons/default/index.theme` is not parsed, it is written by lxqt-config-appearance. The default cursor theme and size is read from X11 calls. There are lots of `index.theme` parsed by liblxqt-config-cursor. It is reasonable to parse `~/.icons/default/index.theme`, too... **Done**

The liblxqt-config-cursor/cfgfile.cpp parses `index.theme` files. It could be done by QSettings. The loadCfgFile function should be removed and use QSettings instead. But I think that it must be in another pull request.

tsujan commented:
> The liblxqt-config-cursor/cfgfile.cpp parses index.theme files. It could be done by QSettings.

Maybe, when that function was written, `QSettings` couldn't handle spaces in keys or values. As far as I tested, it can in Linux.

Or there was another reason?

selairi commented:
> > The liblxqt-config-cursor/cfgfile.cpp parses index.theme files. It could be done by QSettings.
> 
> Maybe, when that function was written, `QSettings` couldn't handle spaces in keys or values. As far as I tested, it can in Linux.
 
The problems one by one, please. First, I would prefer merge the actual PR. Then, a new issue should be opened about QSettings and liblxqt-config-cursor.

tsujan commented:
> The problems one by one, please.

I didn't mean it should be fixed/changed right now or even before the next release.

I'll review the recent changes to this PR tonight.

tsujan commented on liblxqt-config-cursor/crtheme.cpp:
```diff
@@ -265,13 +275,33 @@ bool applyTheme(const XCursorThemeData &theme, int cursorSize)
-        XFixesChangeCursorByName(QX11Info::display(), cursor, QFile::encodeName(name).constData());
+
+        if(QGuiApplication::platformName() == QStringLiteral("xcb"))
+            XFixesChangeCursorByName(QX11Info::display(), cursor, QFile::encodeName(name).constData());
```
in `applyTheme()`, there are too many xcb checks (especially inside the loop).

If your goal is returning `true` from this function when the platform is not xcb, IMHO, it's better to do it once, before the comment `// Sets default cursor size`:

```c++
    if(QGuiApplication::platformName() != QStringLiteral("xcb"))
        return true;
```

selairi replied:
OK

tsujan commented on lxqt-config-appearance/configothertoolkits.cpp:
```diff
@@ -61,20 +64,39 @@ gtk-menu-images = %4
+ * GTK just doesn't intend to support image menu items. They're considered bad practice in modern UI. 
+ */
+
+ConfigOtherToolKits::ConfigOtherToolKits(LXQt::Settings *settings,  LXQt::Settings *configAppearanceSettings, QWidget *parent) : QObject(parent)
```
Would you please tell why you changed `QObject *parent` to `QWidget *parent`?

selairi replied:
Sorry. I changed but I don't remember why. QObject is better than QWidget here. I will restore QObject.

tsujan commented on lxqt-config-appearance/configothertoolkits.cpp:
```diff
@@ -189,6 +218,24 @@ void ConfigOtherToolKits::setXSettingsConfig()
+        QStringList args = QProcess::splitCommand(QStringView(command));
+        bool ok = QProcess::startDetached(QStringLiteral("gsettings"), args);
+        if(!ok)
+            QMessageBox::critical((QWidget*) parent(), tr("Error"), tr("Error: gsettings can not be run"));
```
1. If `gsettings` is not found, the error message needs to be shown only once and the loop should be ended. Here, it's shown with each loop cycle.
 2. A small typo: it's "Error: gsettings cannot be run" (you wrote "can not").

selairi replied:
Ok. Fixed.

tsujan commented on lxqt-config-appearance/main.cpp:
```diff
@@ -104,21 +107,27 @@ int main (int argc, char **argv)
+    QObject::connect(cursorPage, &SelectWnd::settingsChanged, dialog, [dialog, cursorChanged] {
+            dialog->enableButton(QDialogButtonBox::Apply, true);
+            *cursorChanged = true;
+            });
```
I'm not sure whether it's necessary to show a message but, in case it is:

Instead of this, wouldn't it be better if `SelectWnd` had a small method to tell whether the cursor was changed or not? I think a private boolean and a public method that gets it would be enough.

tsujan replied:
OK, I tested under X11 (our main goal) and the cursor was changed on the fly. The message isn't needed.

If it's needed under Wayland, please limit it to Wayland by adding the above-mentioned method to `SelectWnd`. However, I think it's better not to show any message under Wayland either.

selairi replied:
> OK, I tested under X11 (our main target) and the cursor was changed on the fly. The message isn't needed.

Open a LXQt session with openbox as window manager. Then change the cursor style. Finally, put your cursor over an openbox title bar. Openbox keeps the previous cursor theme.
I have installed mousepad in my computer to test GTK3 themes, and mousepad doesn't change cursor theme.

I have just found this: If you see on top of SelectWnd, there is this text: "LXQt needs restart after this change". I will remove the dialog.

Thanks for your review.

tsujan replied:
> Openbox keeps the previous cursor theme.

I tested with KWin; I don't use Openbox. But you're right.

tsujan reviewed (approved):
GTM
